src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -97,25 +233,231 @@ public void Method1()
+using System;
+unsafe class Test
+{
+    static void Method1()
+    {
+        int i = 0;
+        for (var p = stackalloc int[3] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+
+    static void Method2()
+    {
+        int i = 0;
+        for (var p = stackalloc int[ ] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+
+    static void Method3()
+    {
+        int i = 0;
+        for (var p = stackalloc    [ ] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+    
+    public static void Main()
+    {
+        Method1();
+        Method2();
+        Method3();
+    }
+}", TestOptions.UnsafeReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: "123123123", verify: Verification.Fails);
+        }
+
+        [Fact]
+        public void TestFor_Span()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+class Test
+{
+    static void Method1()
+    {
+        int i = 0;
+        for (Span<int> p = stackalloc int[3] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+
+    static void Method2()
+    {
+        int i = 0;
+        for (Span<int> p = stackalloc int[ ] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+
+    static void Method3()
+    {
+        int i = 0;
+        for (Span<int> p = stackalloc    [ ] { 1, 2, 3 }; i < 3; i++)
+            Console.Write(p[i]);
+    }
+    
+    public static void Main()
+    {
+        Method1();
+        Method2();
+        Method3();
+    }
+}", TestOptions.DebugExe);
+
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TestAwait_Pointer()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System.Threading.Tasks;
+unsafe class Test
+{
+    async void M()
+    {
+        var p = stackalloc int[await Task.FromResult(1)] { await Task.FromResult(2) };
+    }
+}", TestOptions.UnsafeReleaseDll);
+
+            comp.VerifyDiagnostics(
+                // (7,32): error CS4004: Cannot await in an unsafe context
+                //         var p = stackalloc int[await Task.FromResult(1)] { await Task.FromResult(2) };
+                Diagnostic(ErrorCode.ERR_AwaitInUnsafeContext, "await Task.FromResult(1)").WithLocation(7, 32),
+                // (7,60): error CS4004: Cannot await in an unsafe context
+                //         var p = stackalloc int[await Task.FromResult(1)] { await Task.FromResult(2) };
+                Diagnostic(ErrorCode.ERR_AwaitInUnsafeContext, "await Task.FromResult(2)").WithLocation(7, 60)
+                );
+        }
+
+        [Fact]
+        public void TestAwait_Span()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+using System.Threading.Tasks;
+class Test
+{
+    async void M()
+    {
+        Span<int> p = stackalloc int[await Task.FromResult(1)] { await Task.FromResult(2) };
```

Comment:
nice. I thought we might have some work with stack-spilling of initializers/size because of `await` . It looks like there is no scenario here that would not be a compile error though due to constraints of async methods and unsafe/span.